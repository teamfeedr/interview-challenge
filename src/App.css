.wrapper {
  padding: 50px 0;
  font-family: 'Nunito', sans-serif;
}

.menu-summary {
  background: #f8f8f8;
  padding: 30px 0;
}

.dietary {
  font-size: 10pt;
  border-radius: 100%;
  background: #a0c9cd;
  box-shadow: inset 0 0 2px rgba(0, 0, 0, 0.1);
  width: 26px;
  height: 26px;
  text-align: center;
  font-weight: bold;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  margin: 5px;
  color: #fff;
}

.menu-summary-left {
  font-weight: bold;
}
.menu-summary-right {
  text-align: right;
}

.menu-builder {
  margin: 50px 0;
}

.item-picker {
  background: #f8f8f8;
  padding: 20px;
  list-style: none;
  margin: 0;
  height: 500px;
  overflow: scroll;
  border-radius: 5px;
}

.menu-preview {
  padding: 20px;
  list-style: none;
  margin: 0;
}

li.item {
  border-radius: 5px;
  background: #fff;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.1);
  padding: 10px;
  margin-bottom: 20px;
  position: relative;
}

li.item h2 {
  font-size: 12pt;
}

li.item p {
  margin: 0;
}

li.item-disabled {
  background: #999;
}

button.remove-item {
  position: absolute;
  top: -13px;
  right: -13px;
  -webkit-appearance: none;
  border: 2px solid #000;
  background: #fff;
  font-weight: bold;
  width: 26px;
  height: 26px;
  line-height: 0px;
  border-radius: 100%;
  padding: 0;
  cursor: pointer;
}

button.remove-item:focus {
  outline: none;
}
