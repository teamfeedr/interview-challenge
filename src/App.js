import React from 'react';
import items from './items';
import MenuBuilder from './MenuBuilder';
import './App.css';

export default () => <MenuBuilder items={items} />;
