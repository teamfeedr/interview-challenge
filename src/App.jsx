import React, { useState } from 'react';
import PropTypes from 'prop-types';

import Menu from './components/Menu/Menu';
import Summary from './components/Summary/Summary';
import Preview from './components/Preview/Preview';
import WithDietaryMap from './components/HOC/WithDietaryMap/WithDietaryMap';
import WithSelectHandler from './components/HOC/WithSelectHandler/WithSelectHandler';
import WithDeselectHandler from './components/HOC/WithDeselectHandler/WithDeselectHandler';

import './App.css';

const App = ({ items }) => {
  const [selectedItems, setSelectedItems] = useState([]);

  return (
    <div className="wrapper">
      <WithDietaryMap items={selectedItems} WrappedComponent={Summary} />

      <div className="container menu-builder">
        <div className="row">
          <div className="col-4">
            <WithSelectHandler
              selectedItemsState={selectedItems}
              updateSelectedItemsState={setSelectedItems}
              WrappedComponent={Menu}
              items={items} />
          </div>

          <div className="col-8">
            <WithDeselectHandler
              selectedItemsState={selectedItems}
              updateSelectedItemsState={setSelectedItems}
              selectedItems={selectedItems}
              WrappedComponent={Preview}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

App.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired
}

export default App;