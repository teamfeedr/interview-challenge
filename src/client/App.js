import React, { useEffect, useState } from 'react';
import { useFetch } from 'use-http';

import ItemPicker from './ItemPicker';
import MenuPreview from './MenuPreview';
import MenuSummary from './MenuSummary';

import './App.css';

export default () => {
  const { get, loading, error } = useFetch();

  const [items, setItems] = useState([])
  const [selectedIds, setSelectedIds] = useState([]);

  useEffect(() => {
    get('/api/items').then((data) => {
      setItems(data?.items);
    });
  }, []);

  const selectItem = (itemId) => setSelectedIds(selectedIds.concat(itemId));

  const removeItem = (itemId) => {
    setSelectedIds(selectedIds.filter((id) => id !== itemId));
  };

  const selectedItems = items.filter((item) => selectedIds.includes(item.id))

  if (loading) {
    return 'Loading...';
  }

  if (error) {
    return 'Something went wrong, please try reloading.';
  }

  return (
    <div className="wrapper">
      <MenuSummary selectedItems={selectedItems} />
      <div className="container menu-builder">
        <div className="row">
          <div className="col-4">
            <ItemPicker
              items={items}
              selectItem={selectItem}
              selectedItems={selectedIds}
            />
          </div>
          <div className="col-8">
            <h2>Menu preview</h2>
            <MenuPreview
              items={selectedItems}
              removeItem={removeItem}
            />
          </div>
        </div>
      </div>
    </div>
  );
};
