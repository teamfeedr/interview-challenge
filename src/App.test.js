import React from "react";
import App from "./App";

import { shallow } from "enzyme";

describe("the App", () => {
  it("renders", () => {
    const div = shallow(<App items={[]} />);
    expect(div).toMatchSnapshot();
  });
});
