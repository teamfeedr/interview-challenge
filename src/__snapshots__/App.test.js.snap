// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "items": Array [
        Object {
          "dietaries": Array [
            "v",
            "ve",
            "df",
            "gf",
            "n!",
          ],
          "id": 1001,
          "name": "Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
            "rsf",
          ],
          "id": 1002,
          "name": "Hake & Smoky Chickpeas, Brown Rice & Quinoa, Roasted Roots",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
            "v",
            "ve",
            "n!",
          ],
          "id": 1003,
          "name": "Dill & Swiss Chard Potato Cakes, Summer Tabbouleh & Roasted Roots",
        },
        Object {
          "dietaries": Array [
            "df",
            "gf",
            "rsf",
            "n!",
          ],
          "id": 1004,
          "name": "Hake & Smoky Chickpeas, Herby Potatoes & Turmeric Satay Broccoli",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
          ],
          "id": 1005,
          "name": "Cuban Beef, Brown Rice & Quinoa, Green Pepper & Butterbean Salad",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
            "rsf",
            "n!",
          ],
          "id": 1006,
          "name": "Thai Chicken, Brown Rice & Quinoa, Lemon & Cashew Greens",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
            "n!",
          ],
          "id": 1007,
          "name": "Cuban Beef, Herby Potatoes & Turmeric Satay Broccoli",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
            "rsf",
            "n!",
          ],
          "id": 1008,
          "name": "Thai Chicken, Brown Rice & Quinoa & Turmeric Satay Broccolli",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
          ],
          "id": 1009,
          "name": "Mackerel Salad Box",
        },
        Object {
          "dietaries": Array [
            "ve",
            "v",
            "gf",
            "df",
          ],
          "id": 10010,
          "name": "Pasta Salad Box",
        },
        Object {
          "dietaries": Array [],
          "id": 10011,
          "name": "Tuna Salad Box",
        },
        Object {
          "dietaries": Array [
            "gf",
            "v",
          ],
          "id": 10012,
          "name": "Coconut Blondie with Dark Chocolate",
        },
        Object {
          "dietaries": Array [
            "gf",
            "df",
          ],
          "id": 10013,
          "name": "Cinnamon Swirl Muffin",
        },
        Object {
          "dietaries": Array [
            "df",
            "gf",
            "n!",
            "v",
          ],
          "id": 10014,
          "name": "Apple, Date and Walnut Muffin",
        },
        Object {
          "dietaries": Array [
            "v",
            "df",
            "gf",
          ],
          "id": 10015,
          "name": "Blueberry and Banana Muffin",
        },
        Object {
          "dietaries": Array [
            "df",
            "ve",
            "gf",
          ],
          "id": 10016,
          "name": "Ombar Chocolate Bar - Dark 72%",
        },
        Object {
          "dietaries": Array [
            "ve",
            "gf",
          ],
          "id": 10017,
          "name": "Ombar Bar - Milk",
        },
        Object {
          "dietaries": Array [
            "v",
            "gf",
          ],
          "id": 10018,
          "name": "Dark Chocolate Brownie",
        },
        Object {
          "dietaries": Array [
            "ve",
            "df",
            "gf",
          ],
          "id": 10019,
          "name": "Mangajo Pomegranate",
        },
        Object {
          "dietaries": Array [
            "ve",
            "v",
            "gf",
          ],
          "id": 10020,
          "name": "Mangajo Acai",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "items": Array [
          Object {
            "dietaries": Array [
              "v",
              "ve",
              "df",
              "gf",
              "n!",
            ],
            "id": 1001,
            "name": "Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
              "rsf",
            ],
            "id": 1002,
            "name": "Hake & Smoky Chickpeas, Brown Rice & Quinoa, Roasted Roots",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
              "v",
              "ve",
              "n!",
            ],
            "id": 1003,
            "name": "Dill & Swiss Chard Potato Cakes, Summer Tabbouleh & Roasted Roots",
          },
          Object {
            "dietaries": Array [
              "df",
              "gf",
              "rsf",
              "n!",
            ],
            "id": 1004,
            "name": "Hake & Smoky Chickpeas, Herby Potatoes & Turmeric Satay Broccoli",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
            ],
            "id": 1005,
            "name": "Cuban Beef, Brown Rice & Quinoa, Green Pepper & Butterbean Salad",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
              "rsf",
              "n!",
            ],
            "id": 1006,
            "name": "Thai Chicken, Brown Rice & Quinoa, Lemon & Cashew Greens",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
              "n!",
            ],
            "id": 1007,
            "name": "Cuban Beef, Herby Potatoes & Turmeric Satay Broccoli",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
              "rsf",
              "n!",
            ],
            "id": 1008,
            "name": "Thai Chicken, Brown Rice & Quinoa & Turmeric Satay Broccolli",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
            ],
            "id": 1009,
            "name": "Mackerel Salad Box",
          },
          Object {
            "dietaries": Array [
              "ve",
              "v",
              "gf",
              "df",
            ],
            "id": 10010,
            "name": "Pasta Salad Box",
          },
          Object {
            "dietaries": Array [],
            "id": 10011,
            "name": "Tuna Salad Box",
          },
          Object {
            "dietaries": Array [
              "gf",
              "v",
            ],
            "id": 10012,
            "name": "Coconut Blondie with Dark Chocolate",
          },
          Object {
            "dietaries": Array [
              "gf",
              "df",
            ],
            "id": 10013,
            "name": "Cinnamon Swirl Muffin",
          },
          Object {
            "dietaries": Array [
              "df",
              "gf",
              "n!",
              "v",
            ],
            "id": 10014,
            "name": "Apple, Date and Walnut Muffin",
          },
          Object {
            "dietaries": Array [
              "v",
              "df",
              "gf",
            ],
            "id": 10015,
            "name": "Blueberry and Banana Muffin",
          },
          Object {
            "dietaries": Array [
              "df",
              "ve",
              "gf",
            ],
            "id": 10016,
            "name": "Ombar Chocolate Bar - Dark 72%",
          },
          Object {
            "dietaries": Array [
              "ve",
              "gf",
            ],
            "id": 10017,
            "name": "Ombar Bar - Milk",
          },
          Object {
            "dietaries": Array [
              "v",
              "gf",
            ],
            "id": 10018,
            "name": "Dark Chocolate Brownie",
          },
          Object {
            "dietaries": Array [
              "ve",
              "df",
              "gf",
            ],
            "id": 10019,
            "name": "Mangajo Pomegranate",
          },
          Object {
            "dietaries": Array [
              "ve",
              "v",
              "gf",
            ],
            "id": 10020,
            "name": "Mangajo Acai",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
